hydra-node/src/Hydra/Chain/Direct.hs

```diff
@@ -197,36 +163,7 @@ withDirectChain tracer networkId iocp socketPath keyPair party cardanoKeys point
             -- connection with the server's tip. So any transaction submitted
             -- before that tip will be missed.
             threadDelay 2
-            action $
-              Chain
-                { postTx = \tx -> do
-                    traceWith tracer $ ToPost tx
-                    timeHandle <- queryTimeHandle networkId socketPath
-                    -- XXX(SN): 'finalizeTx' retries until a payment utxo is
-                    -- found. See haddock for details
-                    response <-
-                      timeoutThrowAfter (NoPaymentInput @Tx) 10 $
-                        -- FIXME (MB): 'cardanoKeys' should really not be here. They
-                        -- are only required for the init transaction and ought to
-                        -- come from the _client_ and be part of the init request
-                        -- altogether. This goes in the direction of 'dynamic
-                        -- heads' where participants aren't known upfront but
-                        -- provided via the API. Ultimately, an init request from
-                        -- a client would contain all the details needed to
-                        -- establish connection to the other peers and to
-                        -- bootstrap the init transaction.
-                        -- For now, we bear with it and keep the static keys in
-                        -- context.
-                        fromPostChainTx timeHandle cardanoKeys wallet headState tx
-                          >>= finalizeTx wallet headState . toLedgerTx
-                          >>= \vtx -> do
-                            response <- newEmptyTMVar
-                            writeTQueue queue (vtx, response)
-                            return response
-                    atomically (takeTMVar response) >>= \case
-                      Nothing -> pure ()
-                      Just err -> throwIO err
-                }
+            action $ mkChain tracer (queryTimeHandle networkId socketPath) cardanoKeys wallet headState queue
```

Comment:
What a weird interface we have there with the `queue`. I see it's unchanged in your refactor and it seems to be enqueuing the tx for submission and blocking on a response by the node. Hence, it should be the responsibility of this module.. not the `Handlers` one?

I would propose to not use the `Chain` interface in this `Handlers` module, but have a function which returns the `Tx` instead of the current `Chain.postTx` interface and do the enqueuing / waiting for errors in this module up here. i.e. something like

```suggestion
            action $ 
              Chain
                { postTx = \tx -> do
                  vtx <- Handlers.createTx tracer (queryTimeHandle networkId socketPath) cardanoKeys wallet headState tx
                  response <- newEmptyTMVar
                  writeTQueue queue (vtx, response)
                  atomically (takeTMVar response) >>= \case
                    Nothing -> pure ()
                    Just err -> throwIO err
                }
```

Reply:
I would kind of agree but for the fact I really wanted to do this refactoring to be able to test this very function in an isolated way, or rather independently from the underlying network machinery.

Reply:
Ah so you mean it's important to you that a `Chain` interface is exposed by the `Handlers` module? I can live with this change overall, but want to state that while it moves things a bit out of the way, it certainly is not easier to understand with the `TMVar` in the `TQueue`.